openedx/openedx-events
# feat: [FC-0074] add support for annotated python dicts as avro map type

## Description

This PR supports Python dict types mapped to Avro Map type for avro schema generation, aiming to add support to a broader list of events payloads. This was previously attempted by mapping Python dicts -> records but this approach considers Python dicts -> maps to avoid conflicts with data attributes -> record mapping, also by using maps we avoid this kind of errors when we don't know the content of dictionaries:

![image](https://github.com/user-attachments/assets/dcf82480-1a91-4e20-adf6-95d8653bdf0f)

~This PR also refactors forum-related events so they can be sent through the event bus. Because of backward compatibility, those changes should be studied in a different PR but are here for testing simplicity.~ Here is the additional PR: https://github.com/openedx/openedx-events/pull/447

## Supporting information

This PR addresses https://github.com/openedx/openedx-events/pull/428#discussion_r1876885613

## Testing instructions

To test with the event bus:
1. Install this branch into your environment (MJG/add-support-for-dicts-forum). The forum events from that branch support dicts with type annotations, so they're a good candidate for testing.
2. Install an event bus backend like the [redis implementation](https://github.com/openedx/event-bus-redis/) in the LMS
3. Setup the event bus with tutor [tutor_installation.rst](https://github.com/openedx/event-bus-redis/blob/main/docs/). Installing discovery is not required, but you can do so and test the block lifecycle events to check that the event bus is working as expected.
4. Create a tutor plugin with this configuration so the event is produced to redis:
```
from tutor import hooks

redis_config = [
    "EVENT_BUS_PRODUCER = 'edx_event_bus_redis.create_producer'",
    "EVENT_BUS_REDIS_CONNECTION_URL = 'redis://@redis:6379/'",
    "EVENT_BUS_TOPIC_PREFIX = 'dev'",
    "EVENT_BUS_CONSUMER = 'edx_event_bus_redis.RedisEventConsumer'",
]

hooks.Filters.ENV_PATCHES.add_item(
    (
        "openedx-common-settings", "\n".join(redis_config)
    )
)

event_bus_config = """
EVENT_BUS_PRODUCER_CONFIG = {
    'org.openedx.learning.forum.thread.created.v1': {
         'forum-thread-created': {'event_key_field': 'thread.id', 'enabled': True},
     },
     'org.openedx.learning.forum.thread.response.created.v1': {
         'forum-thread-response-created': {'event_key_field': 'thread.id', 'enabled': True},
     },
     'org.openedx.learning.forum.thread.response.comment.created.v1': {
         'forum-thread-response-comment-created': {'event_key_field': 'thread.id', 'enabled': True},
     },
}
"""

hooks.Filters.ENV_PATCHES.add_item(
    (
        "openedx-common-settings", event_bus_config
    )
)
```
You should see a log similar to this in the LMS:
![image](https://github.com/user-attachments/assets/f86a035f-f046-4c3e-8b4d-1ed9eea7735a)

## Deadline

None

## Other information

This was previously attempted here: but using record type instead of the map type: https://github.com/openedx/openedx-events/pull/232

## Checklists

Check off if complete *or* not applicable:

**Merge Checklist:**

**Post Merge:**
      finished.
